# Sowing Success: How Machine Learning Helps Farmers Select the Best Crops

![Farmer in a field](farmer_in_a_field.jpg)

Measuring essential soil metrics such as nitrogen, phosphorous, potassium levels, and pH value is an important aspect of assessing soil condition. However, it can be an expensive and time-consuming process, which can cause farmers to prioritize which metrics to measure based on their budget constraints.

Farmers have various options when it comes to deciding which crop to plant each season. Their primary objective is to maximize the yield of their crops, taking into account different factors. One crucial factor that affects crop growth is the condition of the soil in the field, which can be assessed by measuring basic elements such as nitrogen and potassium levels. Each crop has an ideal soil condition that ensures optimal growth and maximum yield.

A farmer reached out to you as a machine learning expert for assistance in selecting the best crop for his field. They've provided you with a dataset called `soil_measures.csv`, which contains:

- `"N"`: Nitrogen content ratio in the soil
- `"P"`: Phosphorous content ratio in the soil
- `"K"`: Potassium content ratio in the soil
- `"pH"` value of the soil
- `"crop"`: categorical values that contain various crops (target variable).

Each row in this dataset represents various measures of the soil in a particular field. Based on these measurements, the crop specified in the `"crop"` column is the optimal choice for that field.  

In this project, you will build multi-class classification models to predict the type of `"crop"` and identify the single most importance feature for predictive performance.

From this information, create a variable called `"best_predictive_feature"` which is a dictionary containing the best predictive feature name as a key and the evaluation score (for the metric you chose) as the value.

In [2]:
# All required libraries are imported here for you.
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.pipeline import Pipeline

# Load the dataset
crops = pd.read_csv("soil_measures.csv")

# Display first few rows
crops.head()

,N,P,K,ph,crop
0,90,42,43,6.502985,rice
1,85,58,41,7.038096,rice
2,60,55,44,7.840207,rice
3,74,35,40,6.980401,rice
4,78,42,42,7.628473,rice


### Check for null values

In [3]:
crops.isna().sum().sort_values()

N       0
P       0
K       0
ph      0
crop    0
dtype: int64

### Check how many possible values does crop have

In [4]:
crops['crop'].value_counts()

crop
rice           100
maize          100
chickpea       100
kidneybeans    100
pigeonpeas     100
mothbeans      100
mungbean       100
blackgram      100
lentil         100
pomegranate    100
banana         100
mango          100
grapes         100
watermelon     100
muskmelon      100
apple          100
orange         100
papaya         100
coconut        100
cotton         100
jute           100
coffee         100
Name: count, dtype: int64

### Model Creation

In [ ]:
# Store the feature names in an array
features = ["N","P","K","ph"]

# Instantiate variables to store the best performing feature
best_feature = ""
best_accuracy = 0

# Get the target variable
y = crops["crop"].values

# Transform non-numerical labels to numerical 
le = LabelEncoder()
y_encoded = le.fit_transform(y)

# Instantiate the pipeline steps
steps=[('scaler',StandardScaler()),('logreg',LogisticRegression())]

# Create the pipeline
pipeline = Pipeline(steps)

# A for loop that computes the accuracy for each feature
for feature in features:
    
    # Get the values of feature of interest
    X = crops[feature].values
    X = X.reshape(-1,1)
    
    # Split the data
    X_train, X_test, y_train, y_test = train_test_split(X,y_encoded,test_size = 0.3,random_state=42)
    
    # Train the model 
    model = pipeline.fit(X_train,y_train)
    
    # Predict using the test set
    y_pred = model.predict(X_test)
    
    # Calculate the accuracy
    accuracy = metrics.accuracy_score(y_test,y_pred)
    
    # Logic to store the best performing feature
    if(accuracy > best_accuracy):
        best_feature = feature
        best_accuracy = accuracy 
        
    print("Feature Used: " + feature)
    print(f"Accuracy: {accuracy:.2f}\n")
    
print("Best Feature: " + best_feature)
print("Accuracy: " + str(best_accuracy))


Feature Used: N
Accuracy: 0.15

Feature Used: P
Accuracy: 0.17

Feature Used: K
Accuracy: 0.24

Feature Used: ph
Accuracy: 0.11

Best Feature: K
Accuracy: 0.2409090909090909


### Store the result in a dictionary


In [6]:
best_predictive_feature = {best_feature:best_accuracy}

### V2: Using all Features in one model

In [7]:
steps=[('scaler',StandardScaler()),('logreg',LogisticRegression())]
pipeline = Pipeline(steps)


X = crops.drop("crop",axis=1).values
y = crops["crop"].values
le = LabelEncoder()
y_encoded = le.fit_transform(y)

X_train, X_test, y_train, y_test = train_test_split(X,y_encoded,test_size = 0.3,random_state=42)
model = pipeline.fit(X_train,y_train)

y_pred = model.predict(X_test)

accuracy = metrics.accuracy_score(y_test,y_pred)    
print(f"Model Accuracy: {accuracy:.2f}\n")



Model Accuracy: 0.65



Since the task specified to determine the single best-performing feature, the best performing feature is the 'K' feature with an accuracy of 0.24

However, it is worth noting that using more features can lead to better accuracy. This is validated by the 0.65 accuracy displayed by the model that used all features.